### ISSUE TITLE
Huge performance regression during JIT code generation with ILGenerator on .NET 6.0 (Windows x64)

### ISSUE THREAD

**Issue by kpreisser** [dotnet/runtime#68953]
### Description

Hi, after upgrading from .NET 5.0.16 to .NET 6.0.4, we encountered a seemingly huge performance regression when using `System.Reflection.Emit.ILGenerator` (using `DynamicMethod`). It occurs in [Jurassic](https://github.com/paulbartrum/jurassic), a .NET JavaScript runtime that compiles JavaScript code to .NET IL code as lightweight functions.
**Note:** The performance issue occurs while the JIT is **generating**  the native code, not in the resulting native code generated by the JIT.

With the reproduction project described below, when runnig with .NET 5.0.16 (on Windows 10 Version 21H2 x64), I get the following execution times depending on the build configuration:

Debug: 60 ms
Release: 130 ms

When executing it with .NET 6.0.4, I get the following times:

Debug: 54 ms
Release: 4300 ms

Notice that for Release, with .NET 6.0.4 the execution time is about 30x as high as with .NET 5.0.16.
I also tested it on Ubuntu 20.04, and got similar results.

### Reproduction Steps

I created a repro project that uses [Jurassic](https://github.com/paulbartrum/jurassic) to compile JavaScript code in .NET code with `System.Reflection.Emit.ILGenerator` using a `DynamicMethod` (unfortunatly I'm not sure how to recreate the issue without Jurassic, as I'mt not very experienced with `ILGenerator`).
The `DynamicMethod` is created in [`MethodGenerator.cs`, line 228](https://github.com/kpreisser/jurassic/blob/720b1617c18ef0059b48bd10eda8edda4522669e/Jurassic/Compiler/MethodGenerator/MethodGenerator.cs#L228).

Jurassic will generate 3 `DynamicMethods` for the three script functions in the JavaScript code (one for the global method, one for the function `doThings`, and one for the anonymous funciton expression in `partitions.sort`.

To reproduce the issue:

- Clone https://github.com/kpreisser/jurassic.git and switch to branch **`reproJitPerformanceRegression`**.
- `cd` into `ReproJitPerfRegression`, then run the following commands:
- `dotnet run -f net5.0 -c Release`
- `dotnet run -f net6.0 -c Debug`
- `dotnet run -f net6.0 -c Release`

### Expected behavior

When running `dotnet run -f net6.0 -c Release`, the time shown after the first `Executed script after:` should be similar to the time reported for `net5.0`, which is about 130 ms. E.g.:

```
Framework: .NET 5.0.16
Compiled script after: 00:00:00.0844951
BEFORE Function
START Function!
AFTER Function, execution took 133 ms
Executed script after: 00:00:00.1492869
BEFORE Function
START Function!
AFTER Function, execution took 0 ms
Executed script after: 00:00:00.0013050
```

### Actual behavior

When running `dotnet run -f net6.0 -c Release`, the time shown after the first `Executed script after:` is about 30x higher, e.g. 4300 ms. E.g.:

```
Framework: .NET 6.0.4
Compiled script after: 00:00:00.1040805
BEFORE Function
START Function!
AFTER Function, execution took 4234 ms
Executed script after: 00:00:04.2561898
BEFORE Function
START Function!
AFTER Function, execution took 0 ms
Executed script after: 00:00:00.0013538
```

**Note:** As you can see from the logging, the huge delay (4200 ms) only occurs the **first time** executing the script function, so it seems to occur while the JIT generates the native code, not in the actual native code generated by the JIT.

### Regression?

Yes, with .NET 5.0.16, as noted above, the execution time is reasonable (around 130 ms).

### Known Workarounds

Compile the assembly that uses `ILGenerator` without Code Optimizations enabled (e.g. as `Debug` instead of `Release`), but that may affect performance of the native code generated by the JIT.

### Configuration

* .NET 6.0.4, .NET SDK 6.0.202
* OS: Windows 10 Pro Version 21H2 Build 19044, x64
* Arch: x64
* It probably also occurs on other OSes, I also reproduced it on Ubuntu 20.04 x64.

**Comment by jakobbotsch** [comment]
Might potentially be the same as #66076 and fixed in main already by #66259, though I haven't verified.

**Comment by AndyAyersMS** [comment]
Is there some prerequisite install needed? When I build the repo I get errors:
```
C:\Users\andya\.nuget\packages\microsoft.build.tasks.git\1.0.0-beta2-18618-05\build\Microsoft.Build.Tasks.Git.targets(20,5): error MSB4018: The "Microsoft.Build.Tasks.Git.LocateRepository" task failed unexpectedly. [D:\bugs\r68953\jurassic\Jurassic\Jurassic.csproj]
C:\Users\andya\.nuget\packages\microsoft.build.tasks.git\1.0.0-beta2-18618-05\build\Microsoft.Build.Tasks.Git.targets(20,5): error MSB4018: System.TypeInitializationException: The type initializer for 'Microsoft.Build.Tasks.Git.TaskImplementation' threw an exception. [D:\bugs\r68953\jurassic\Jurassic\Jurassic.csproj]
```

**Comment by AndyAyersMS** [comment]
I got things working by updating Microsoft.SourceLink.GitHub to Version 1.1.1.

**Comment by AndyAyersMS** [comment]
> Might potentially be the same as #66076 and fixed in main already by #66259, though I haven't verified.

Yes, looks like the same issue, lots of time spent under `optRedundantBranch`.

<img width="854" alt="image" src="https://user-images.githubusercontent.com/10121823/167176858-09e6af36-ddbc-46d6-a40f-81f5e8a6f92f.png">

**Comment by AndyAyersMS** [comment]
With 7.0 the time is more reasonable (though not as fast at 5.0).

5.0: 122
6.0: 3967
7.0: 482

```
D:\bugs\r68953\jurassic\ReproJitPerfRegression>dotnet run -f net7.0 -c Release
Framework: .NET 7.0.0-preview.3.22175.4
Compiled script after: 00:00:00.0920905
BEFORE Function
START Function!
AFTER Function, execution took 482 ms
Executed script after: 00:00:00.5014486
BEFORE Function
START Function!
AFTER Function, execution took 0 ms
Executed script after: 00:00:00.0010012

D:\bugs\r68953\jurassic\ReproJitPerfRegression>dotnet run -f net6.0 -c Release
Framework: .NET 6.0.4
Compiled script after: 00:00:00.0706652
BEFORE Function
START Function!
AFTER Function, execution took 3967 ms
Executed script after: 00:00:03.9826453
BEFORE Function
START Function!
AFTER Function, execution took 0 ms
Executed script after: 00:00:00.0009346

D:\bugs\r68953\jurassic\ReproJitPerfRegression>dotnet run -f net5.0 -c Release
Framework: .NET 5.0.16
Compiled script after: 00:00:00.0755503
BEFORE Function
START Function!
AFTER Function, execution took 122 ms
Executed script after: 00:00:00.1419613
BEFORE Function
START Function!
AFTER Function, execution took 1 ms
Executed script after: 00:00:00.0009057
```

**Comment by AndyAyersMS** [comment]
There might be ways to alter your IL generation to avoid this but it might take a bit of experimentation to hit on the best approach. The pathological case is when your method contains a long chain of similar branches, eg repeated if-then-elses with the same predicate. If you can vary the predicate and/or reduce the length of the chain you should be able to mitigate the problem.

Since this is the second case of this issue we've seen. I'll also propose that we fix this in 6.0 servicing.

**Comment by AndyAyersMS** [comment]
This is fixed in 7.0; will leave issue open until we know if it will be fixed in 6.0 as well.

**Comment by AndyAyersMS** [comment]
Some details on the method that causes trouble:
```
;; DynamicClass::doThings

38821 bytes of IL
3782 basic blocks (when RBO runs)

;; problematic recurring predicate

ldarg.1     
castclass    0x200002A (token can vary)

Because we're optimizing this sequence turns into a dominating null check on arg1. There are around 816 of these.
```

**Comment by kpreisser** [comment]
> Is there some prerequisite install needed? When I build the repo I get errors:

Sorry, I forgit tp push a commit that removed those references. It should work now right away..

> This is fixed in 7.0; will leave issue open until we know if it will be fixed in 6.0 as well.

Thanks, that's good to hear! I would appreciate if this could be backported to 6.0, as we are using that version in a production environment, and don't want to switch the Jurassic assembly to `Debug` since we don't want to miss the code optimizations in the generated native code.

Thanks!

**Comment by AndyAyersMS** [comment]
@kpreisser the fix will be in 6.06 which I think should be released in June. I'll ping you once it is out so you can give it a try.

**Comment by AndyAyersMS** [comment]
This is fixed in main / 7.0 and will be fixed shortly with 6.0.6.

**Comment by kpreisser** [comment]
Thank you! I can confirm that with 6.0.6 the performance is much better than with 6.0.5.
